src/main/java/net/neoforged/neoforge/client/model/data/ModelDataManager.java

```diff
@@ -23,142 +22,130 @@
+    public ModelData getAt(BlockPos pos) {
+        Preconditions.checkArgument(level.isClientSide, "Cannot request model data for server level");
+        long sectionPos = SectionPos.asLong(pos);
+        refreshAt(sectionPos);
+        return modelDataCache.getOrDefault(sectionPos, Long2ObjectMaps.emptyMap()).getOrDefault(pos.asLong(), ModelData.EMPTY);
+    }
 
     /**
-     * The active manager owned by the client's level and operated on the main client thread.
-     * <p>
-     * Users should not be instantiating this themselves unless they know what they're doing.
+     * Snapshot the state of this manager for all sections in the volume specified by the given section coordinates.
+     *
+     * @throws IllegalArgumentException if this is a snapshot and the given region doesn't match the snapshot's region
      */
-    @ApiStatus.Internal
-    public static final class Active extends ModelDataManager {
-        private final Thread owningThread = Thread.currentThread();
-        private final Level level;
-        private final Long2ObjectMap<Set<BlockPos>> needModelDataRefresh = new Long2ObjectOpenHashMap<>();
-        private final Long2ObjectMap<Long2ObjectMap<ModelData>> modelDataCache = new Long2ObjectOpenHashMap<>();
-
-        public Active(Level level) {
-            this.level = level;
-        }
-
-        /**
-         * Request a refresh of the stored data for the given {@link BlockEntity}. The given {@code BlockEntity}
-         * must be in the level owning this manager
-         */
-        public void requestRefresh(BlockEntity blockEntity) {
-            if (isOtherThread()) {
-                throw new UnsupportedOperationException("Cannot request ModelData refresh outside the owning thread: " + owningThread);
-            }
-
-            Preconditions.checkNotNull(blockEntity, "BlockEntity must not be null");
-            Preconditions.checkState(blockEntity.getLevel() == level, "BlockEntity does not belong to the level owning this manager");
-            needModelDataRefresh.computeIfAbsent(SectionPos.asLong(blockEntity.getBlockPos()), $ -> new HashSet<>())
-                    .add(blockEntity.getBlockPos());
-        }
-
-        @Override
-        @Nullable
-        public ModelData getAt(BlockPos pos) {
-            Preconditions.checkArgument(level.isClientSide, "Cannot request model data for server level");
-            long sectionPos = SectionPos.asLong(pos);
-            refreshAt(sectionPos);
-            return modelDataCache.getOrDefault(sectionPos, Long2ObjectMaps.emptyMap()).get(pos.asLong());
-        }
-
-        @Override
-        public ModelData getAtOrEmpty(BlockPos pos) {
-            return Objects.requireNonNullElse(getAt(pos), ModelData.EMPTY);
+    public ModelDataManager.Snapshot snapshotSectionRegion(int sectionMinX, int sectionMinY, int sectionMinZ, int sectionMaxX, int sectionMaxY, int sectionMaxZ) {
```

Comment:
Since `Snapshot` is `Internal`, this method should be too; however, I'm not sure `Snapshot` should be internal. It's just a wrapper around a `Long2ObjectMap`, but so is the return of `Long2ObjectMaps.unmodifiable(map)`. I think the copying code in the constructor of `Snapshot` should be moved to this method, this method should return `@Unmodifiable Long2ObjectMap<ModelData>`, and `Snapshot` should be removed.